modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/servlet/DLCopyEntryServlet.java

```diff
@@ -0,0 +1,288 @@
+			HttpServletResponse httpServletResponse)
+		throws IOException, ServletException {
+
+		try {
+			User user = _portal.getUser(httpServletRequest);
+
+			if ((user == null) || user.isGuestUser()) {
+				throw new PrincipalException.MustBeAuthenticated(
+					StringPool.BLANK);
+			}
+
+			PrincipalThreadLocal.setName(user.getUserId());
+
+			PermissionThreadLocal.setPermissionChecker(
+				_permissionCheckerFactory.create(user));
+
+			super.service(httpServletRequest, httpServletResponse);
+		}
+		catch (IOException | PortalException | ServletException exception) {
+			_log.error(exception);
+
+			_sendResponse(
+				httpServletResponse,
+				HttpServletResponse.SC_INTERNAL_SERVER_ERROR, null, null);
+		}
+	}
+
+	private void _checkDestinationRepository(long repositoryId)
+		throws PortalException {
+
+		Group group = _groupLocalService.fetchGroup(repositoryId);
+
+		if ((group != null) && group.isStaged() && !group.isStagingGroup()) {
+			throw new PortalException(
+				"cannot-copy-folders-to-the-live-version-of-a-group");
+		}
+	}
+
+	private void _copyFileEntry(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException {
+
+		long fileEntryId = ParamUtil.getLong(httpServletRequest, "fileEntryId");
+		long destinationFolderId = ParamUtil.getLong(
+			httpServletRequest, "destinationFolderId");
+		long destinationRepositoryId = ParamUtil.getLong(
+			httpServletRequest, "destinationRepositoryId");
+
+		try {
+			_checkDestinationRepository(destinationRepositoryId);
+
+			_dlAppService.copyFileEntry(
+				fileEntryId, destinationFolderId, destinationRepositoryId,
+				ServiceContextFactory.getInstance(
+					DLFileShortcut.class.getName(), httpServletRequest));
+
+			_sendResponse(
+				httpServletResponse, HttpServletResponse.SC_OK,
+				_language.get(
+					httpServletRequest, "your-request-completed-successfully"),
+				null);
+		}
+		catch (PortalException portalException) {
+			_sendResponse(
+				httpServletResponse, HttpServletResponse.SC_METHOD_NOT_ALLOWED,
+				null, portalException.getMessage());
+		}
+	}
+
+	private void _copyFileShortcut(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException {
+
+		long fileShortcutId = ParamUtil.getLong(
+			httpServletRequest, "fileShortcutId");
+		long destinationFolderId = ParamUtil.getLong(
+			httpServletRequest, "destinationFolderId");
+		long destinationRepositoryId = ParamUtil.getLong(
+			httpServletRequest, "destinationRepositoryId");
+
+		try {
+			_checkDestinationRepository(destinationRepositoryId);
+
+			_dlAppService.copyFileShortcut(
+				fileShortcutId, destinationFolderId, destinationRepositoryId,
+				ServiceContextFactory.getInstance(
+					DLFileShortcut.class.getName(), httpServletRequest));
+
+			_sendResponse(
+				httpServletResponse, HttpServletResponse.SC_OK,
+				_language.get(
+					httpServletRequest, "your-request-completed-successfully"),
```

Comment:
Why are you passing something that doesn't change across different calls (the error message)? Remove that parameter and get the message internally in the `_sendResponse` method.

Reply:
Forget about this. I had misread the code; it is fine as it is.